### ISSUE TITLE
python 3.11.0 fails to build on macOS (OpenSSL build failure)

### ISSUE THREAD

**Issue by minddrive** [pyenv/pyenv#2494]
Platform: macOS 12.6
OS architecture: x86_64 (Intel)
pyenv version: 2.3.5-3-g0726e02e
Python version: 3.11.0
C Compiler information: Apple clang version 14.0.0 (clang-1400.0.29.102)

Building the new release of Python 3.11.0 fails due to a failure for the build of OpenSSL 1.1.1q:

```
Installing openssl-1.1.1q...

BUILD FAILED (OS X 12.6 using python-build 2.3.5-3-g0726e02e)

Inspect or clean up the working tree at /var/folders/d7/04qpwbhs5mg_m5rycyyk_3x40000gp/T/python-build.20221024151009.558
Results logged to /var/folders/d7/04qpwbhs5mg_m5rycyyk_3x40000gp/T/python-build.20221024151009.558.log

Last 10 log lines:
                 apps/libapps.a -lssl test/libtestutil.a -lcrypto
ld: warning: directory not found for option '-L/Users/relayer/.pyenv/versions/3.11.0/lib'
clang  -Iinclude -arch x86_64 -O3 -Wall -D_REENTRANT -DZLIB -DZLIB_SHARED -DNDEBUG -I/Users/relayer/.pyenv/versions/3.11.0/include  -MMD -MF test/v3ext.d.tmp -MT test/v3ext.o -c -o test/v3ext.o test/v3ext.c
test/v3ext.c:201:24: error: implicitly declaring library function 'memcmp' with type 'int (const void *, const void *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
        if (!TEST_true(memcmp(ip1->data, ip2->data, ip1->length) <= 0))
                       ^
test/v3ext.c:201:24: note: include the header <string.h> or explicitly provide a declaration for 'memcmp'
1 error generated.
make[1]: *** [test/v3ext.o] Error 1
make: *** [all] Error 2
```

This is actually a long known issue: https://github.com/openssl/openssl/issues/18720 .  The fix isn't planned until 1.1.1r (likely several months away), but there's a simple patch (a single include line) that can fix this.  Might it be possible to get this added to correct the problem?  (If I knew of the process, I could offer a PR, if needed.)

**Comment by minddrive** [comment]
I wanted to give an update in that I found out that Homebrew is apparently able to build OpenSSL 1.1.1q without issue, and as a result was able to build Python 3.11.0.  With this workaround, I'm uncertain if this issue is as critical, but will let the appropriate people decide.

**Comment by TaiPhamD** [comment]
@minddrive how did you get 3.11 to show up? for some reason my pyenv only sees 3.11.0rc2
```
➜ pyenv --version
pyenv 2.3.5

➜ pyenv install --list | grep 3.11
  3.11.0rc2
  3.11-dev
  miniconda3-4.3.11

```

**Comment by minddrive** [comment]
@TaiPhamD , I ran 'pyenv update' to get the latest updates for available versions, which brought up 3.11.0.

**Comment by AvdN** [comment]
> @TaiPhamD , I ran 'pyenv update' to get the latest updates for available versions, which brought up 3.11.0.

I installed [pyenv-update](https://github.com/pyenv/pyenv-update), to get the update sub-command. Then ran `pyenv update` and got "Already up to date". Still can't see or install 3.11.0.

**Comment by minddrive** [comment]
@AvdN I can't say what's happening, but https://github.com/pyenv/pyenv/blob/master/plugins/python-build/share/python-build/3.11.0 shows the release version has been added, and I've now built it successfully using Homebrew's version of OpenSSL.

**Comment by AvdN** [comment]
> @AvdN I can't say what's happening, but https://github.com/pyenv/pyenv/blob/master/plugins/python-build/share/python-build/3.11.0 shows the release version has been added, and I've now built it successfully using Homebrew's version of OpenSSL.

So did I after installing the HEAD version of pyenv (see the updates in my previous comment)

**Comment by gerbendekkervolue** [comment]
Great that 3.11.0 was added to HEAD so quickly! Anything blocking providing a release from pyenv?

**Comment by boweeb** [comment]
Minor correction

@AvdN -- I think this is what you intended (last two cmds: `brew` --> `pyenv`)?

```bash
brew unlink pyenv
brew install --HEAD pyenv
pyenv install --list | grep 3.11    
pyenv install 3.11.0
```

**Comment by AvdN** [comment]
@boweb Yes you a right, thanks for catching that.

**Comment by billti** [comment]
Had the same issue re not seeing 3.11.0 and ran the above, and I got the below errors :-/. I literally just installed Homebrew and pyenv in the last hour without any issues (after upgrade to Ventura on this new MacBook). I did see homebrew installing the Xcode-select tools (even though I'd already installed them. Maybe it downgraded them?)

(This may well be a homebrew issue and not a pyenv issue, just sharing here in case others try the above).

```
 ~/ brew unlink pyenv 
Unlinking /opt/homebrew/Cellar/pyenv/2.3.5... 5 symlinks removed.
 ~/ brew install --HEAD pyenv
==> Cloning https://github.com/pyenv/pyenv.git
Cloning into '/Users/billti/Library/Caches/Homebrew/pyenv--git'...
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
Error: Your Xcode (14.0.1) is too outdated.
Please update to Xcode 14.1 (or delete it).
Xcode can be updated from the App Store.

Error: Your Command Line Tools (CLT) does not support macOS 13.
It is either outdated or was modified.
Please update your Command Line Tools (CLT) or delete it if no updates are available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install
```

**Comment by TaiPhamD** [comment]
> Had the same issue re not seeing 3.11.0 and ran the above, and I got the below errors :-/. I literally just installed Homebrew and pyenv in the last hour without any issues (after upgrade to Ventura on this new MacBook). I did see homebrew installing the Xcode-select tools (even though I'd already installed them. Maybe it downgraded them?)

I had this issue too but you can resolve it by installing Xcode 14.1 RC. I don't think 14.1 GA is out yet.

**Comment by billti** [comment]
Thanks. I installed Xcode 14.1 RC 2 and got the same error. I tried running `Xcode-select --install` again which told me they were already installed, however I could see the files in `/Library/Developer/CommandLineTools/usr/bin` (clang, etc.) were mostly from August. Downloading and installing the RC2 Command Line Tools pkg from https://developer.apple.com/download/applications/ resolved the issue, and the pyenv install of 3.11.0 went without a hitch (see below) on Ventura. Thanks again!

```
 ~/ pyenv install 3.11.0
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.11.0.tar.xz...
-> https://www.python.org/ftp/python/3.11.0/Python-3.11.0.tar.xz
Installing Python-3.11.0...
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
Installed Python-3.11.0 to /Users/billti/.pyenv/versions/3.11.0
```

**Comment by polydevuk** [comment]
So what's the consensus on the build failure?

**Comment by TaiPhamD** [comment]
> So what's the consensus on the build failure?

- The OP failed the build because of missing openSSL dev lib which he fixed using brew to install openSSL (The pyenv manual actually gives instruction for us to install python build dependencies: see [here](https://github.com/pyenv/pyenv/wiki#suggested-build-environment))

- Then one other user could not build due to recent macOS update to Ventura that required Xcode 14.1 . 

- Other discussions are just to get 3.11 to show up which was solved by using brew install --head pyenv

**Comment by danxmoran** [comment]
At least one person at my company is hitting this same error trying to install Python 2.9 on MacOS 12 - will the fix from #2500 apply to that build as well? Or only to Python 2.11+?

**Comment by native-api** [comment]
@danxmoran try applying the patch to the source of OpenSSL specified in that version's formula. If it fits, you can just copy the patch to the appropriate path. We'd appreciate it if you make a PR.
